cpp/include/raft/neighbors/detail/knn_brute_force.cuh

```diff
@@ -0,0 +1,588 @@
+    raft::linalg::transpose(handle, search, search_row_major.data(), n, D, userStream);
+    search = search_row_major.data();
+  }
+
+  // Make other streams from pool wait on main stream
+  handle.wait_stream_pool_on_stream();
+
+  for (size_t i = 0; i < input.size(); i++) {
+    value_t* out_d_ptr = out_D + (i * k * n);
+    IdxType* out_i_ptr = out_I + (i * k * n);
+
+    auto stream = handle.get_next_usable_stream(i);
+
+    if (k <= 64 && rowMajorQuery == rowMajorIndex && rowMajorQuery == true &&
+        (metric == raft::distance::DistanceType::L2Unexpanded ||
+         metric == raft::distance::DistanceType::L2SqrtUnexpanded ||
+         metric == raft::distance::DistanceType::L2Expanded ||
+         metric == raft::distance::DistanceType::L2SqrtExpanded)) {
+      fusedL2Knn(D,
+                 out_i_ptr,
+                 out_d_ptr,
+                 input[i],
+                 search_items,
+                 sizes[i],
+                 n,
+                 k,
+                 rowMajorIndex,
+                 rowMajorQuery,
+                 stream,
+                 metric);
+
+      // Perform necessary post-processing
+      if (metric == raft::distance::DistanceType::L2SqrtExpanded ||
+          metric == raft::distance::DistanceType::L2SqrtUnexpanded ||
+          metric == raft::distance::DistanceType::LpUnexpanded) {
+        float p = 0.5;  // standard l2
+        if (metric == raft::distance::DistanceType::LpUnexpanded) p = 1.0 / metricArg;
+        raft::linalg::unaryOp<float>(
+          res_D,
+          res_D,
+          n * k,
+          [p] __device__(float input) { return powf(fabsf(input), p); },
+          stream);
+      }
+    } else {
+      switch (metric) {
+        case raft::distance::DistanceType::Haversine:
+          ASSERT(D == 2,
+                 "Haversine distance requires 2 dimensions "
+                 "(latitude / longitude).");
+
+          haversine_knn(out_i_ptr, out_d_ptr, input[i], search_items, sizes[i], n, k, stream);
+          break;
+        default:
+
+          auto index = input[i];
+          rmm::device_uvector<value_t> index_row_major(0, userStream);
+          if (!rowMajorIndex) {
+            index_row_major.resize(sizes[i] * D, userStream);
+            raft::linalg::transpose(handle, index, index_row_major.data(), sizes[i], D, userStream);
+            index = index_row_major.data();
+
+            // Make other streams from pool wait on main stream
+            handle.wait_stream_pool_on_stream();
+          }
+
+          // cosine/correlation are handled by metric processor, use IP distance
+          // for brute force knn call.
+          auto tiled_metric = metric;
+          if (metric == raft::distance::DistanceType::CosineExpanded ||
+              metric == raft::distance::DistanceType::CorrelationExpanded) {
+            tiled_metric = raft::distance::DistanceType::InnerProduct;
+          }
+
+          tiled_brute_force_knn<value_t, IdxType>(handle,
```

Comment:
So for example, above we're doing `handle.get_next_available_stream()` and we're passing that into the `fusedL2NN` to enable it to be asynchronous wrt that stream in the stream pool. But here, we're passing the same handle that was passed into this function, so it's going to once again synchronize w/ the user stream, meaning it's not actually honoring the stream pool at all.

We should be able to create a new handle w/ the `get_next_available_stream()`. Or even better, copy the handle so it doesn't have to create a new cublas handle underneath and set the user stream on it.

Reply:
I'm creating a copy of the handle - and then setting the handle returned from get_next_available_stream on it in https://github.com/rapidsai/raft/pull/1202/commits/9593ae16b527c32d8b46e88fef946c8c0a46a999